core/unit_test/hip/TestHIP_InterOp_Graph.cpp

```diff
@@ -0,0 +1,97 @@
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <filesystem>
+
+#include <TestHIP_Category.hpp>
+#include <TestGraph_helpers.hpp>
+#include <Kokkos_Graph.hpp>
+
+#include <gtest/gtest.h>
+
+namespace {
+
+// This test checks the promises of Kokkos::Graph against its
+// underlying HIP native objects.
+TEST(TEST_CATEGORY, graph_promises_on_native_objects) {
+  auto graph = Kokkos::Experimental::create_graph<Kokkos::HIP>();
+
+  auto root = Kokkos::Impl::GraphAccess::create_root_ref(graph);
+
+  auto graph_ptr_impl =
+      Kokkos::Impl::GraphAccess::get_graph_weak_ptr(root).lock();
+
+  static_assert(std::is_same_v<decltype(graph_ptr_impl)::element_type,
+                               Kokkos::Impl::GraphImpl<Kokkos::HIP>>);
+
+  // Before instantiation, the HIP graph is valid, but the HIP executable
+  // graph is still null.
+  hipGraph_t hip_graph = graph_ptr_impl->get_hip_graph();
+
+  ASSERT_NE(hip_graph, nullptr);
+  ASSERT_EQ(graph_ptr_impl->get_hip_graph_exec(), nullptr);
+
+  // After instantiation, both native objects are valid.
+  graph.instantiate();
+
+  hipGraphExec_t hip_graph_exec = graph_ptr_impl->get_hip_graph_exec();
+
+  ASSERT_EQ(graph_ptr_impl->get_hip_graph(), hip_graph);
+  ASSERT_NE(hip_graph_exec, nullptr);
+
+  // Submission should not affect the underlying objects.
+  graph.submit();
+
+  ASSERT_EQ(graph_ptr_impl->get_hip_graph(), hip_graph);
+  ASSERT_EQ(graph_ptr_impl->get_hip_graph_exec(), hip_graph_exec);
+}
+
+// Use native HIP graph to generate a DOT representation.
+TEST(TEST_CATEGORY, graph_instantiate_and_debug_dot_print) {
+  using view_t = Kokkos::View<int, Kokkos::HIP>;
+
+  const Kokkos::HIP exec{};
+
+  view_t data(Kokkos::view_alloc(exec, "witness"));
+
+  auto graph = Kokkos::Experimental::create_graph(exec);
+
+  auto root = Kokkos::Impl::GraphAccess::create_root_ref(graph);
+
+  root.then_parallel_for(1, SetViewToValueFunctor<Kokkos::HIP, int>{data, 1});
+
+  auto graph_ptr_impl =
+      Kokkos::Impl::GraphAccess::get_graph_weak_ptr(root).lock();
+
+  graph.instantiate();
+
+  size_t num_nodes = 0;
+
+  KOKKOS_IMPL_HIP_SAFE_CALL(
+      hipGraphGetNodes(graph_ptr_impl->get_hip_graph(), nullptr, &num_nodes));
+
+  ASSERT_EQ(num_nodes, 2u);
+
+  const auto dot = std::filesystem::temp_directory_path() / "hip_graph.dot";
+
+  KOKKOS_IMPL_HIP_SAFE_CALL(
+      hipGraphDebugDotPrint(graph_ptr_impl->get_hip_graph(), dot.c_str(),
+                            hipGraphDebugDotFlagsVerbose));
+
+  ASSERT_TRUE(std::filesystem::exists(dot));
```

Comment:
How good is that test? Is a file, possibly empty, created if the graph is invalid?

Reply:
I could check if the file is not empty, but checking the content is harder (and I guesswe don't want to hardcode the expected result, as it depends on pointer addresses and stuff).

Reply:
I am asking because if `hipGraphDebugDotPrint` creates a file no matter what, the test is useless.

Reply:
OK, so I've just written a check for what's the content of the exported `dot`. But the test relies too much on internals (like the threshold for constant vs global memory launch).

So I will revert this.

I think that the test in its current form is indeed not very meaningful. However, noting that the graph has been instantiated just before, this means the backend had a chance to shout if the graph was invalid.

Making a "very meaningful" use of the executable graph would make the test much more complex (for instance I could [change the kernel parameters](https://rocm.docs.amd.com/projects/HIP/en/latest/doxygen/html/group___graph.html#ga5b1918dae65224863b7370e6d4ad3f2a)), so I think just adding a check that the file is not empty would be enough. Do you agree ?